File: scripts/linter/src/rules/test-macro-spacing.ts
import { LintRule, LintContext, LintViolation } from '../types';

// Test macros that need spaces inside parentheses: MACRO( content )
const TEST_MACROS = [
    // Test structure
    'TEST_CASE',
    'SUBCASE',

    // Basic assertions
    'CHECK',
    'CHECK_FALSE',
    'REQUIRE',
    'REQUIRE_FALSE',
    'WARN',
    'WARN_FALSE',

    // Binary comparisons
    'CHECK_EQ',
    'CHECK_NE',
    'CHECK_GT',
    'CHECK_LT',
    'CHECK_GE',
    'CHECK_LE',
    'REQUIRE_EQ',
    'REQUIRE_NE',
    'REQUIRE_GT',
    'REQUIRE_LT',
    'REQUIRE_GE',
    'REQUIRE_LE',
    'WARN_EQ',
    'WARN_NE',
    'WARN_GT',
    'WARN_LT',
    'WARN_GE',
    'WARN_LE',

    // Unary
    'CHECK_UNARY',
    'CHECK_UNARY_FALSE',
    'REQUIRE_UNARY',
    'REQUIRE_UNARY_FALSE',
    'WARN_UNARY',
    'WARN_UNARY_FALSE',

    // Exception handling
    'CHECK_THROWS',
    'CHECK_THROWS_AS',
    'CHECK_THROWS_WITH',
    'CHECK_THROWS_WITH_AS',
    'CHECK_NOTHROW',
    'REQUIRE_THROWS',
    'REQUIRE_THROWS_AS',
    'REQUIRE_THROWS_WITH',
    'REQUIRE_THROWS_WITH_AS',
    'REQUIRE_NOTHROW',
    'WARN_THROWS',
    'WARN_THROWS_AS',
    'WARN_THROWS_WITH',
    'WARN_THROWS_WITH_AS',
    'WARN_NOTHROW',

    // Message variants
    'CHECK_MESSAGE',
    'REQUIRE_MESSAGE',
    'WARN_MESSAGE',

    // Logging
    'INFO',
    'CAPTURE',
    'MESSAGE',
    'FAIL',
    'FAIL_CHECK',
];

// Pattern: MACRO(x - no space after opening paren
// Matches MACRO( followed immediately by a non-space, non-) character
function createNoLeadingSpacePattern(macros: string[]): RegExp {
    const macroPattern = macros.join('|');
    // Match MACRO( followed by non-whitespace (but not just closing paren)
    return new RegExp(`\\b(${macroPattern})\\(([^\\s)])`, 'g');
}

// Pattern: MACRO ( - space before opening paren (should be MACRO()
function createSpaceBeforeParenPattern(macros: string[]): RegExp {
    const macroPattern = macros.join('|');
    return new RegExp(`\\b(${macroPattern})\\s+\\(`, 'g');
}

// Pattern: MACRO(content ) without proper closing - content) instead of content )
// This is trickier - we need to find ) that isn't preceded by space
function checkClosingSpace(line: string, macroName: string, openParenIndex: number): { hasError: boolean; column: number } | null {
    let depth = 1;
    let i = openParenIndex + 1;

    while (i < line.length && depth > 0) {
        const char = line[i];

        // Handle string literals
        if (char === '"' || char === "'") {
            const quote = char;
            i++;
            while (i < line.length && line[i] !== quote) {
                if (line[i] === '\\') i++; // Skip escaped char
                i++;
            }
        } else if (char === '(') {
            depth++;
        } else if (char === ')') {
            depth--;
            if (depth === 0) {
                // Check if there's no space before this )
                const prevChar = i > 0 ? line[i - 1] : ' ';
                if (prevChar !== ' ' && prevChar !== '\t') {
                    return { hasError: true, column: i + 1 };
                }
            }
        }
        i++;
    }

    return null;
}

// Check if a line is a comment
function isComment(line: string): boolean {
    const trimmed = line.trim();
    return trimmed.startsWith('//') || trimmed.startsWith('/*') || trimmed.startsWith('*');
}

export const testMacroSpacingRule: LintRule = {
    name: 'test-macro-spacing',
    description: 'Test macros should have spaces inside parentheses: MACRO( content )',

    check(context: LintContext): LintViolation[] {
        const violations: LintViolation[] = [];
        const noLeadingSpacePattern = createNoLeadingSpacePattern(TEST_MACROS);
        const spaceBeforeParenPattern = createSpaceBeforeParenPattern(TEST_MACROS);

        for (let i = 0; i < context.lines.length; i++) {
            const line = context.lines[i];
            const lineNumber = i + 1;

            // Skip comment lines
            if (isComment(line)) {
                continue;
            }

            // Check for space before opening paren: MACRO ( instead of MACRO(
            spaceBeforeParenPattern.lastIndex = 0;
            let spaceMatch;

            while ((spaceMatch = spaceBeforeParenPattern.exec(line)) !== null) {
                const macroName = spaceMatch[1];
                const matchStart = spaceMatch.index;

                violations.push({
                    rule: 'test-macro-spacing',
                    message: `Unexpected space before '(' in '${macroName}'`,
                    line: lineNumber,
                    column: matchStart + macroName.length + 1,
                    severity: 'error',
                });
            }

            // Check for missing leading space: MACRO(x instead of MACRO( x
            noLeadingSpacePattern.lastIndex = 0;
            let match;

            while ((match = noLeadingSpacePattern.exec(line)) !== null) {
                const macroName = match[1];
                const matchStart = match.index;

                violations.push({
                    rule: 'test-macro-spacing',
                    message: `Missing space after '(' in '${macroName}'`,
                    line: lineNumber,
                    column: matchStart + macroName.length + 2,
                    severity: 'error',
                });
            }

            // Check for missing trailing space before )
            for (const macroName of TEST_MACROS) {
                const macroPattern = new RegExp(`\\b${macroName}\\(`, 'g');
                let macroMatch;

                while ((macroMatch = macroPattern.exec(line)) !== null) {
                    const openParenIndex = macroMatch.index + macroName.length;
                    const closingCheck = checkClosingSpace(line, macroName, openParenIndex);

                    if (closingCheck?.hasError) {
                        violations.push({
                            rule: 'test-macro-spacing',
                            message: `Missing space before ')' in '${macroName}'`,
                            line: lineNumber,
                            column: closingCheck.column,
                            severity: 'error',
                        });
                    }
                }
            }
        }

        return violations;
    },
};

File: scripts/linter/rules/test_macro_spacing.cpp
#include "../linter.hpp"

#include <cctype>

namespace wisdom_linter
{
namespace
{
    auto isComment( std::string_view line ) -> bool
    {
        size_t start = line.find_first_not_of( " \t" );
        if ( start == std::string_view::npos )
        {
            return false;
        }
        std::string_view trimmed = line.substr( start );
        return trimmed.starts_with( "//" ) || trimmed.starts_with( "/*" ) ||
               trimmed.starts_with( "*" );
    }

    class TestMacroSpacingRule : public Rule
    {
    public:
        [[nodiscard]] auto name() const -> std::string_view override
        {
            return "test-macro-spacing";
        }

        [[nodiscard]] auto description() const -> std::string_view override
        {
            return "Test macros should have spaces inside parentheses: MACRO( content )";
        }

        [[nodiscard]] auto check( const LintContext& context ) const
            -> std::vector<LintViolation> override
        {
            std::vector<LintViolation> violations;

            static const std::vector<std::string> test_macros = {
                "TEST_CASE", "SUBCASE", "CHECK", "CHECK_FALSE", "REQUIRE", "REQUIRE_FALSE",
                "WARN", "WARN_FALSE", "INFO", "CAPTURE", "GENERATE", "SECTION",
                "CHECK_EQ", "CHECK_NE", "CHECK_GT", "CHECK_LT", "CHECK_GE", "CHECK_LE",
                "REQUIRE_EQ", "REQUIRE_NE", "REQUIRE_GT", "REQUIRE_LT", "REQUIRE_GE", "REQUIRE_LE",
                "WARN_EQ", "WARN_NE", "WARN_GT", "WARN_LT", "WARN_GE", "WARN_LE",
                "CHECK_UNARY", "CHECK_UNARY_FALSE",
                "REQUIRE_UNARY", "REQUIRE_UNARY_FALSE",
                "WARN_UNARY", "WARN_UNARY_FALSE",
                "REQUIRE_THROWS", "REQUIRE_THROWS_AS", "REQUIRE_THROWS_WITH",
                "REQUIRE_THROWS_WITH_AS", "REQUIRE_NOTHROW",
                "CHECK_THROWS", "CHECK_THROWS_AS", "CHECK_THROWS_WITH",
                "CHECK_THROWS_WITH_AS", "CHECK_NOTHROW",
                "WARN_THROWS", "WARN_THROWS_AS", "WARN_THROWS_WITH", "WARN_THROWS_WITH_AS",
                "WARN_NOTHROW",
                "CHECK_MESSAGE", "REQUIRE_MESSAGE", "WARN_MESSAGE",
                "MESSAGE", "FAIL", "FAIL_CHECK",
            };

            for ( size_t i = 0; i < context.lines.size(); ++i )
            {
                const auto& line = context.lines[i];
                int line_number = static_cast<int>( i + 1 );

                if ( isComment( line ) )
                {
                    continue;
                }

                for ( const auto& macro : test_macros )
                {
                    size_t pos = 0;
                    while ( ( pos = line.find( macro, pos ) ) != std::string::npos )
                    {
                        if ( pos > 0 && ( std::isalnum( line[pos - 1] ) || line[pos - 1] == '_' ) )
                        {
                            ++pos;
                            continue;
                        }

                        size_t after_macro = pos + macro.size();
                        size_t open_paren = after_macro;
                        bool has_space_before_paren = false;

                        while ( open_paren < line.size() &&
                                ( line[open_paren] == ' ' || line[open_paren] == '\t' ) )
                        {
                            has_space_before_paren = true;
                            ++open_paren;
                        }

                        if ( open_paren >= line.size() || line[open_paren] != '(' )
                        {
                            ++pos;
                            continue;
                        }

                        if ( has_space_before_paren )
                        {
                            violations.push_back( LintViolation {
                                std::string { name() },
                                "Unexpected space before '(' in '" + macro + "'",
                                line_number,
                                static_cast<int>( open_paren + 1 ),
                                Severity::Error,
                            } );
                        }
                        if ( open_paren + 1 < line.size() && line[open_paren + 1] != ' ' &&
                             line[open_paren + 1] != ')' )
                        {
                            violations.push_back( LintViolation {
                                std::string { name() },
                                "Missing space after '(' in '" + macro + "'",
                                line_number,
                                static_cast<int>( open_paren + 2 ),
                                Severity::Error,
                            } );
                        }

                        int depth = 1;
                        size_t j = open_paren + 1;
                        bool in_string = false;
                        char string_char = 0;

                        while ( j < line.size() && depth > 0 )
                        {
                            char c = line[j];

                            if ( !in_string && ( c == '"' || c == '\'' ) )
                            {
                                in_string = true;
                                string_char = c;
                            }
                            else if ( in_string && c == string_char && ( j == 0 || line[j - 1] != '\\' ) )
                            {
                                in_string = false;
                            }
                            else if ( !in_string )
                            {
                                if ( c == '(' )
                                {
                                    ++depth;
                                }
                                else if ( c == ')' )
                                {
                                    --depth;
                                    if ( depth == 0 && j > 0 && line[j - 1] != ' ' &&
                                         line[j - 1] != '\t' )
                                    {
                                        violations.push_back( LintViolation {
                                            std::string { name() },
                                            "Missing space before ')' in '" + macro + "'",
                                            line_number,
                                            static_cast<int>( j + 1 ),
                                            Severity::Error,
                                        } );
                                    }
                                }
                            }
                            ++j;
                        }

                        pos = open_paren + 1;
                    }
                }
            }

            return violations;
        }
    };
} // namespace

auto createTestMacroSpacingRule() -> std::shared_ptr<Rule>
{
    return std::make_shared<TestMacroSpacingRule>();
}

} // namespace wisdom_linter
